From 2b1c8ba0c5ee3acc2f9f221d517114c28b2c8bf9 Mon Sep 17 00:00:00 2001
From: Hood Chatham <roberthoodchatham@gmail.com>
Date: Sat, 2 Apr 2022 11:13:37 -0700
Subject: [PATCH] loadDynamicLibrary flapack

We are using CLAPACK for our LAPACK, but CLAPACK only supports
LAPACK v3.2 since starting in v3.3 LAPACK started adding fortran
code from Fortran standards more recent that 1990 and f2c only
supports Fortran '90. Scipy uses some functions that were added
in LAPACK v3.4, but all of these are Fortran 77 compliant and so
we *can* f2c them. In scipy/meta.yaml we inject these into
`lapack_extras.f` which gets built into `_flapack.so`.

Why do we do it this way? It was the first thing I tried that
worked and I was so exhausted that I didn't have the energy to be
bothered by the hack. Ideally we would copy these into our CLAPACK
but CLAPACK is the result of hand modified f2c output. (In fact
this is how f2c was always intended to be used, the output is not
good C code by itself.) We are automating the patches that are
necessary in `_f2c_fixes.py`, but these don't get applied to CLAPACK.

Anyways, the issue is that CLAPACK is loaded as a "shared library"
which means that we call `loadDynamicLibrary` on it so that all of
its symbols are globally scoped (as opposed to needing to manually
look them up with `dlsym`). Scipy is not loaded this way. But we
need the `lapack_extras.f` symbols to have this global visibility.
So we have to call loadDynamicLibrary on it, hence this patch.
---
 scipy/__init__.py | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/scipy/__init__.py b/scipy/__init__.py
index 8c23069f9..4fa852612 100644
--- a/scipy/__init__.py
+++ b/scipy/__init__.py
@@ -158,3 +158,10 @@ else:
 
     # This makes "from scipy import fft" return scipy.fft, not np.fft
     del fft
+
+    import sysconfig
+    ext_suffix = sysconfig.get_config_var("EXT_SUFFIX")
+    from pyodide_js._module import loadDynamicLibrary
+    loadDynamicLibrary(f'/lib/python3.10/site-packages/scipy/linalg/_flapack{ext_suffix}')
+    del sysconfig
+    del ext_suffix
-- 
2.25.1

